{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48a8ad5b-56be-427e-9076-c0ba15c3989e",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Counts DNA Nucleotides\n",
    "\n",
    "def countNA(x):\n",
    "    A=0\n",
    "    C=0\n",
    "    G=0\n",
    "    T=0\n",
    "    for i in range(len(x)):\n",
    "        if x[i]==\"A\":\n",
    "            A+=1\n",
    "        elif x[i]==\"C\":\n",
    "            C+=1\n",
    "        elif x[i]==\"G\":\n",
    "            G+=1\n",
    "        else:\n",
    "            T+=1\n",
    "    return (A,\" \", C,\" \", G, \" \", T)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "926063b1-1a10-4c21-9727-83b0fa707a8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "##RNA Encoding of a DNA strand\n",
    "\n",
    "def RNA(x):\n",
    "    return(x.replace(\"T\",\"U\"))\n",
    "           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a537696c-d018-408d-83c6-902ad553292b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Returns the complement of a DNA strand of nucleotides\n",
    "\n",
    "def DNAc(x):\n",
    "    \n",
    "    # Turns strand into an array of 0,1,2,3\n",
    "    def translate(x):\n",
    "        T=[]\n",
    "        for i in range(len(x)):\n",
    "            \n",
    "            # These numbers are chosen such that adding 2 to any nucleotide (mod 4) gives the complement nucleotide\n",
    "            if x[i]==\"A\":\n",
    "                T.append(0)\n",
    "            elif x[i]==\"C\":\n",
    "                T.append(3)\n",
    "            elif x[i]==\"T\":\n",
    "                T.append(2)\n",
    "            elif x[i]==\"G\":\n",
    "                T.append(1)\n",
    "        return(T)\n",
    "    \n",
    "    # Turns an array with values 0,1,2,3 into a DNA strand\n",
    "    def translateInv(x):\n",
    "        T=\"\"\n",
    "        for i in range(len(x)):\n",
    "            if x[i]==0:\n",
    "                T+=\"A\"\n",
    "            elif x[i]==3:\n",
    "                T+=\"C\"\n",
    "            elif x[i]==2:\n",
    "                T+=\"T\"\n",
    "            elif x[i]==1:\n",
    "                T+=\"G\"\n",
    "        return(T)\n",
    "    \n",
    "    # Reverse the strand\n",
    "    r=x[::-1]\n",
    "    \n",
    "    # Convert to an array\n",
    "    R=translate(r)\n",
    "    \n",
    "    # Finally to obtain the complement nucleotides add 2 to each component mod 4 and make it a string\n",
    "    return(translateInv(list(map(lambda l : (l + 2)%4, R))))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09808c39-7147-4177-874e-06201bd5dd0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Solves the Fibbonacci rabbit problem using a reccurence relation\n",
    "\n",
    "def Rabbit(n,k):\n",
    "    R=[1,1]\n",
    "    for i in range(2,n):\n",
    "        R.append(k*R[i-2]+R[i-1])\n",
    "    return(R[n-1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3ed2f37-c357-4985-8853-942892c18308",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculates the GC content of some amount of labelled RNA strands and returns the ID of the strand with highest content\n",
    "\n",
    "def GCcontent(x,file):\n",
    "    \n",
    "    # Reads a file containing an amount of labelled RNA strands \n",
    "    x=open(file,\"r\")\n",
    "    lines=x.readlines()\n",
    "    ID=[]\n",
    "    GC=[]\n",
    "    length=0\n",
    "    temp=0\n",
    "    for i in range(len(x)):\n",
    "        \n",
    "        # Checks if the string has terminated and a new string began\n",
    "        if x[i][0]==\">\":\n",
    "            if length>0:\n",
    "                GC.append(temp/length)\n",
    "            ID.append(x[i][1:-1])\n",
    "            length=0\n",
    "            temp=0\n",
    "            \n",
    "        else:\n",
    "            length+=len(x[i])-1\n",
    "            for j in range(len(x[i])):\n",
    "                if x[i][j]==\"G\" or x[i][j]==\"C\":\n",
    "                    temp+=1 \n",
    "            if i==len(x)-1:\n",
    "                GC.append(temp/(length+1))\n",
    "    M=max(GC)\n",
    "    return(ID[GC.index(M)]+\"\\n\"+str(int(M*10**8)/10**6))\n",
    "      \n",
    "print(GCcontent(lines))       \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c8da020-ed70-416b-9909-14fe22ba339b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rosalind_0424\n",
      "52.168525\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0657230-6890-42f6-8a6e-1bce77428154",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Calculates the Hamming distance between two strings\n",
    "\n",
    "def HammingDist(s,t):\n",
    "    temp=0\n",
    "    for i in range(len(s)):\n",
    "        if s[i]!=t[i]:\n",
    "            temp+=1\n",
    "    return temp\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdbe660a-0bbe-41a5-a834-26b986648e4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Opening and processing a text file to make a dictionary for \n",
    "# converting RNA codons to their corresponding amino acids\n",
    "\n",
    "Dictionary=open(\"rosalind_Dictionary.txt\",\"r\")\n",
    "Dic=Dictionary.readlines()\n",
    "for i in range(len(Dic)-1):\n",
    "    Dic[i]=Dic[i][:-1]\n",
    "\n",
    "D={}\n",
    "b=[]\n",
    "for i in range(len(Dic)):\n",
    "    line_data = Dic[i]\n",
    "    a = line_data.split(\" \")\n",
    "    for j in range(len(a)):\n",
    "        if a[j]==\"\":\n",
    "            continue\n",
    "        else:\n",
    "            b.append(a[j])\n",
    "\n",
    "for i in range(int(len(b)/2)):\n",
    "    D[b[2*i]]=b[2*i+1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee6d29e6-22ac-4046-be03-788bbe093c05",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Takes an RNA strand and converts it to a sequence of amino acids (protein)\n",
    "\n",
    "def RNAtoProtein(x):\n",
    "    Protein=''\n",
    "    for i in range(int(len(x)/3)):\n",
    "        temp=x[3*i]+x[3*i+1]+x[3*i+2]\n",
    "        #stops if temp is a stop codon\n",
    "        if D[temp]==\"Stop\":\n",
    "            return(Protein)\n",
    "            break\n",
    "        else:\n",
    "            #otherwise adds the cooresponding amino acid\n",
    "            Protein+=D[temp]\n",
    "\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2654c9f2-7ed8-4bc4-8937-13f4b203171d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Takes a string s and find all motifs t and returns their positions in the string \n",
    "\n",
    "def stringmotif(s,t):\n",
    "    temp=\"\"\n",
    "    P=\"\"\n",
    "    for i in range(len(s)):\n",
    "        if t[0]==s[i]:\n",
    "            for j in range(1,len(t)):\n",
    "                if i+len(t)+1>len(s):\n",
    "                    return(P)\n",
    "                elif s[i+j]==t[j]:\n",
    "                    temp=str(i+1)+\" \"\n",
    "                else:\n",
    "                    temp=\"\"\n",
    "                    break\n",
    "        else:\n",
    "            continue\n",
    "        P=P+temp\n",
    "    return(P)\n",
    "stringmotif(\"GAGGCGCGGAGGCGCCCCCGCCCTCCCGAGGCGCGTAACGAGGCGCAGGAGGCGCGTTTTTAAAAGAGGCGCGAGGCGCCTGAGGCGCACCAGAGGCGCGAGGCGCGAGGCGCGAGGCGCATATTGAGGCGCATCCGAGGCGCCGGGAGGCGCGAGGCGCTCGAGGCGCTGAGGCGCGAGGCGCAGGAGGCGCGAGGCGCATTGAGGCGCTACCCAGAGGCGCGAGGCGCATACGGAGGCGCTCGAGGCGCTCTGAGTGAGGCGCGAGAGGCGCGAGGCGCCTGAGGCGCGAGGCGCTGTTCGCGGAAGAGGCGCCTGCAGGGACCGCAAGAGGCGCCAGAGGCGCTGAGGCGCCAGAGGCGCCGCGAGGCGCAGGTGAGGCGCCGAGGCGCGAGAGGCGCCTGAGGCGCCGAGGCGCGAGGCGCCCTTGAGGCGCCAGGAGGCGCAGGCCGGGTGAGGCGCGCGAGGCGCCACAGAGGCGCCACAGAGGCGCGAGGCGCGAGGCGCTTGAGGCGCGCGAGGCGCGAGGCGCATCGAGGCGCCGAACTATTGAGGCGCCCGAAAAGAGGCGCAGTGCGGACTTGAGGCGCGTGAGGCGCCTGAGGCGCGAGGCGCACTAGAGGCGCTGAGGCGCTGGAGGCGCGGAGGCGCTGAGGCGCTGGAGGCGCGGATAGGAGGCGCTATGAGGCGCGGAGGCGCGAGGCGCGAGGCGCTTGCGAGGCGCAGAGGCGCTCGAGGCGCGAGACCGAGGCGCGGAGGCGCAGGCGAGGCGCGCCCAGGAGGCGCGAGGCGCGGAGGCGCGAGGCGCGAGGCGCGAGGCGCCTGTGGAGGCGCCCGTCCCGAGGCGCCTCGTCTGCTGAGGCGCTGGAGGCGCCCTGGACGTGCTTCAGAGGCGCTGAGGCGCAGAGGCGCACGTATCTGGAGGCGCCTCGAGGCGCCTGAGGCGC\",\"GAGGCGCGA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c69cc481-a91a-482b-951e-b3c1e343f385",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper function that reads FASTA format\n",
    "def readFASTA(x):     \n",
    "    ID=[]\n",
    "    strings=[]\n",
    "    length=0\n",
    "    Dict={}\n",
    "    temp=\"\"\n",
    "    for i in range(len(x)):\n",
    "        if x[i][0]==\">\":\n",
    "            if length>0:\n",
    "                strings.append(temp)\n",
    "            ID.append(x[i][1:-1])\n",
    "            length=0\n",
    "            temp=\"\"\n",
    "        else:\n",
    "            length+=len(x[i])-1\n",
    "            if i==len(x)-1:\n",
    "                temp+=x[i][:-1+(len(x[i])+1)*int(x[i][len(x[i])-1]!=\"\\n\")]\n",
    "                strings.append(temp)\n",
    "            temp+=x[i][:-1]\n",
    "\n",
    "    for i in range(len(ID)):\n",
    "        Dict[ID[i]]=strings[i]\n",
    "    return(Dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0950872e-d18d-4216-b0d6-6f45c561802d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Takes a list of DNA strings in FAST format and returns their 3-overlap graph's adjacency list\n",
    "def Overlap(data):\n",
    "    dic=readFASTA(data)\n",
    "    Keys=\"\"\n",
    "    for key in dic:\n",
    "        for otherkey in dic:\n",
    "            if key==otherkey:\n",
    "                continue\n",
    "            elif dic[key][-3:]==dic[otherkey][:3]:\n",
    "                Keys=Keys+str(key)+ \" \" +str(otherkey)+\"\\n\"\n",
    "    return(Keys)   \n",
    "\n",
    "x=open(\"rosalind_grph.txt\",\"r\").readlines()\n",
    "Overlap(x)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78706b7c-a2b0-47f9-b577-50ca14229bc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Boolean to find the glycos protein motif \n",
    "def IsGlycos(x):\n",
    "    if len(x)!=4:\n",
    "        return(print(\"Error: your input is not 4 proteins long\"))\n",
    "    elif x[0]==\"N\" and x[1]!=\"P\" and (x[2]==\"S\" or x[2]==\"T\") and x[3]!=\"P\":\n",
    "        return True\n",
    "    else:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6214187c-d349-4a39-81ef-efa15d2a01eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Takes a list of proteins in FASTA format and returns the positions of \n",
    "# all glycos motifs\n",
    "\n",
    "def GlycosMotif(x):\n",
    "    contain=[]\n",
    "    out=\"\"\n",
    "    Proteins=readFASTA(x)\n",
    "    for p in Proteins:\n",
    "        temp=\"\"\n",
    "        for i in range(len(Proteins[p])-3):\n",
    "            if IsGlycos(Proteins[p][i:i+4]):\n",
    "                contain.append(p)\n",
    "                if temp==\"\":\n",
    "                    temp+=str(p)+\"\\n\"+str(i+1)+\" \"\n",
    "                else:\n",
    "                    temp+=str(i+1)+\" \"\n",
    "            else:\n",
    "                continue\n",
    "        if p in contain:\n",
    "            out+=temp+\"\\n\"\n",
    "        else:\n",
    "            continue\n",
    "    \n",
    "    return(out)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0ec6ec6-fa25-4305-b99f-049b11375206",
   "metadata": {},
   "outputs": [],
   "source": [
    "y=open(\"Test.txt\",\"r\").readlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd9f919d-c412-4d2e-8d35-0cbf6891f89d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Modified version of the rabbit problem where the rabbits die after m months\n",
    "\n",
    "def MortFibRabbit(n,m):\n",
    "    B=[1]\n",
    "    M=[0]\n",
    "    for i in range(1,m):\n",
    "        M.append(B[i-1]+M[i-1])\n",
    "        B.append(M[i-1])\n",
    "    for i in range(m,n):\n",
    "        M.append(B[i-1]+M[i-1]-B[i-m])\n",
    "        B.append(M[i-1])\n",
    "    return(B[n-1]+M[n-1])\n",
    "\n",
    "MortFibRabbit(97,18)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "589b05f3-76fc-4b97-b04b-6a1248a22a8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Opening a file of RNA strands and processing \n",
    "\n",
    "y=readFASTA(open(\"rosalind_splc (2).txt\",\"r\").readlines())\n",
    "rna=RNA(y[list(y.keys())[0]])\n",
    "y.pop(list(y.keys())[0])\n",
    "lengths=[]\n",
    "\n",
    "\n",
    "for p in y: \n",
    "    lengths.append(len(y[p]))\n",
    "M=min(lengths)\n",
    "\n",
    "# Takes strands of RNA and removes introns and concatenates exons \n",
    "\n",
    "def RNAsplice(x,start,out):\n",
    "    if start+M-1>len(x)-1:\n",
    "        out+=x[start:]\n",
    "        return(RNAtoProtein(out))\n",
    "    else:\n",
    "        for p in y:\n",
    "            if x[start:start+len(y[p])]==RNA(y[p]):\n",
    "                return(RNAsplice(x,start+len(y[p]),out))\n",
    "        out+=x[start]\n",
    "        return(RNAsplice(x,start+1,out))\n",
    "        \n",
    "\n",
    "print(RNAsplice(rna,0,\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06944124-c685-4a71-936a-339f902bbb64",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Incomplete function that should calculate probabilites using Mendel's law and conditional probability\n",
    "def IPRB(k,m,n):\n",
    "    T=k+m+n\n",
    "    p1=k/T\n",
    "    p2=m*k/(T*(T-1))\n",
    "    p3=0.75*m*(m-1)/(T*(T-1))\n",
    "    p4=0.5*m*n/(T*(T-1))\n",
    "    return(p1+p2+p3+p4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52038528-d09e-490d-8826-42eecd5a57c3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
